crates/biome_js_analyze/src/semantic_analyzers/nursery/no_global_assign.rs

```diff
@@ -0,0 +1,141 @@
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// const a = 1;
+    /// let b = 1;
+    /// b = 100;
+    ///
+    /// onload = function() {};
+    ///
+    /// /*global a:writable*/
+    /// a = 100;
+    /// ```
+    ///
+    pub(crate) NoGlobalAssign {
+        version: "next",
+        name: "noGlobalAssign",
+        recommended: false,
+    }
+}
+
+impl Rule for NoGlobalAssign {
+    type Query = Semantic<AnyJsExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let expr = match node {
+            AnyJsExpression::JsAssignmentExpression(expr) => expr,
+            _ => return None,
+        };
+
+        let left_expr = expr.left().ok()?;
+        let right_expr = expr.right().ok()?;
+        let left = left_expr.text();
+        let right_text = right_expr.text();
+
+        let is_function_call = match right_expr {
+            AnyJsExpression::JsFunctionExpression(expr) => expr.function_token().is_ok(),
+            _ => false,
+        };
+
+        let assignment = left_expr.as_any_js_assignment()?;
+        let token = assignment
+            .as_js_identifier_assignment()?
+            .name_token()
+            .ok()?;
+        let comment = token.text();
+        let search_term = format!("global {}:readonly", left);
+
+        let global_text = if NODE.contains(&left.as_str())
+            || BROWSER.contains(&left.as_str())
+            || BUILTIN.contains(&left.as_str())
+            || ES_2021.contains(&left.as_str())
+            || comment.contains(&search_term)
+        {
+            left
+        } else {
+            return None;
+        };
+
+        if !global_text.is_empty() && !right_text.is_empty() && !is_function_call {
+            return Some(());
+        }
+
+        None
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        let expr = match node {
+            AnyJsExpression::JsAssignmentExpression(expr) => expr,
+            _ => return None,
+        };
+        let left_expr = expr.left().ok()?;
+        let left = left_expr.text();
+        let is_native_objects = NODE.contains(&left.as_str())
+            || BROWSER.contains(&left.as_str())
+            || BUILTIN.contains(&left.as_str())
+            || ES_2021.contains(&left.as_str());
+
+        let target = if is_native_objects {
+            "native object"
+        } else {
+            "global variable"
+        };
+
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "Assignment to " {target} ": "  <Emphasis>{left}</Emphasis>
+                },
+            )
+            .note(markup! {
+                "Global variables are not allowed to be reassigned."
+            }),
+        )
```

Comment:
We should explain why it is discouraged.
```suggestion
        Some(
            RuleDiagnostic::new(
                rule_category!(),
                node.range(),
                markup! {
                    "A global variable should not be reassigned"
                },
            )
            .note(markup! {
                "Assigning to a global variable can override essential functionality."
            }),
        )
```